### PR TITLE
MC-23915: Wrong currency symbol in creditmemo_grid & sales_order_view > creditmemo grid for subtotal & Shipping and Handling fee

### LINKED ISSUE #22662: Wrong currency symbol in creditmemo_grid & sales_order_view > creditmemo grid  for subtotal & Shipping and Handling fee
**Preconditions (*)**
0. Magento CE 2.2.8

1.Magento 2.4-develop
2.Create second website with store and store view
3.Set another currency for second website(in my case its Ukrainian Hryvnia)
image
4.Go the Stores-->Currency Rates and set value for UAH = 2 
image

**Steps to reproduce (*)**
1. Create order in a store view with different currency as configured on default store view
2. Go to Backend --> Sales--->Orders and make an invoice and credit memo 
3. Check Subtotal and Shipping and Handling columns on Order page grid and Credit Memos pages grid.

 **Expected result (*)**
1. Display the Subtotal, Shipping & Handling with correct currency sign.

**Actual result (*)**

Wrong currency sign for Subtotal, Shipping & Handling Fee on Order and Credit memos pages grid(correct currency only for Grand Total(Purchased)):

- Order Grid
image

- CreditMemo Grid
image

### PR DESCRIPTION
**Pull request by engcom-Echo**
### Fixed Issues (if relevant)

1. Fixes magento/magento2#22662

### Manual testing scenarios (*)

See : magento/magento2#22662

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by engcom-Echo** [comment]
@magento run all tests

File: app/code/Magento/Sales/Ui/Component/Listing/Column/PurchasedPrice.php [hunk 1]
```diff
@@ -28,6 +29,11 @@ class PurchasedPrice extends Column
      */
     private $currency;
 
+    /**
+     * @var OrderRepositoryInterface
+     */
+    private $order;
```
Comment (ihor-sviziev) [inline app/code/Magento/Sales/Ui/Component/Listing/Column/PurchasedPrice.php]:
```suggestion
    private $orderRepository;
```

File: app/code/Magento/Sales/Ui/Component/Listing/Column/PurchasedPrice.php [hunk 2]
```diff
@@ -62,7 +72,9 @@ public function prepareDataSource(array $dataSource)
     {
         if (isset($dataSource['data']['items'])) {
             foreach ($dataSource['data']['items'] as & $item) {
-                $currencyCode = isset($item['order_currency_code']) ? $item['order_currency_code'] : null;
+                $currencyCode = isset($item['order_currency_code'])
+                    ? $item['order_currency_code']
+                    : $this->order->get($item['order_id'])->getOrderCurrencyCode();
```
Comment (ihor-sviziev) [inline app/code/Magento/Sales/Ui/Component/Listing/Column/PurchasedPrice.php]: 1. Looks like here we can use `??` here

```suggestion
                $currencyCode = $item['order_currency_code'] ?? $this->order->get($item['order_id'])->getOrderCurrencyCode();
```

2. Loading objects in the loop is a really heavy operation and not recommended. Can we load all needed orders at once? Or maybe we can do that when preparing dataSource?

File: app/code/Magento/Sales/Ui/Component/Listing/Column/PurchasedPrice.php [hunk 3]
```diff
@@ -37,18 +43,22 @@ class PurchasedPrice extends Column
...
         array $components = [],
         array $data = [],
-        Currency $currency = null
+        Currency $currency = null,
+        OrderRepositoryInterface $order = null
     ) {
         $this->priceFormatter = $priceFormatter;
         $this->currency = $currency ?: \Magento\Framework\App\ObjectManager::getInstance()
             ->create(Currency::class);
+        $this->order = $order ?: \Magento\Framework\App\ObjectManager::getInstance()
```
Comment (ihor-sviziev) [inline app/code/Magento/Sales/Ui/Component/Listing/Column/PurchasedPrice.php]:
```suggestion
        $this->orderRepository = $orderRepository ?? \Magento\Framework\App\ObjectManager::getInstance()
```

File: app/code/Magento/Sales/Test/Unit/Ui/Component/Listing/Column/PurchasedPriceTest.php [hunk 4]
```diff
@@ -28,6 +29,11 @@ class PurchasedPriceTest extends TestCase
      */
     protected $currencyMock;
 
+    /**
+     * @var OrderRepositoryInterface|MockObject
+     */
+    protected $orderMock;
```
Comment (ihor-sviziev) [inline app/code/Magento/Sales/Test/Unit/Ui/Component/Listing/Column/PurchasedPriceTest.php]:
```suggestion
    private $orderRepositoryMock;
```

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

File: app/code/Magento/Sales/Ui/Component/Listing/Column/PurchasedPrice.php (hunk 2 above)
Comment (engcom-Echo) [reply app/code/Magento/Sales/Ui/Component/Listing/Column/PurchasedPrice.php]: 1. Thanks.
2. Will do

**Comment by engcom-Echo** [comment]
@magento run all tests

File: app/code/Magento/Sales/Test/Unit/Ui/Component/Listing/Column/PurchasedPriceTest.php [hunk 5]
```diff
@@ -21,12 +22,17 @@ class PurchasedPriceTest extends TestCase
...
     /**
      * @var Currency|MockObject
      */
-    protected $currencyMock;
+    private $currencyMock;
+
+    /**
+     * @var OrderRepositoryInterface|MockObject
+     */
+    private $orderRepository;
```
Comment (ihor-sviziev) [inline app/code/Magento/Sales/Test/Unit/Ui/Component/Listing/Column/PurchasedPriceTest.php]:
```suggestion
    private $orderRepositoryMock;
```

File: app/code/Magento/Sales/Test/Unit/Ui/Component/Listing/Column/PurchasedPriceTest.php [hunk 6]
```diff
@@ -41,31 +47,60 @@ protected function setUp(): void
...
+            ->setMethods(['getList','get','delete','save','getOrderCurrencyCode'])
+            ->disableOriginalConstructor()
+            ->getMock();
         $this->model = $objectManager->getObject(
             PurchasedPrice::class,
-            ['currency' => $this->currencyMock, 'context' => $contextMock]
+            [
+                'currency' => $this->currencyMock,
+                'context' => $contextMock,
+                'order' => $this->orderRepository,
```
Comment (ihor-sviziev) [inline app/code/Magento/Sales/Test/Unit/Ui/Component/Listing/Column/PurchasedPriceTest.php]:
```suggestion
                'orderRepository' => $this->orderRepositoryMock,
```

File: app/code/Magento/Sales/Test/Unit/Ui/Component/Listing/Column/PurchasedPriceTest.php [hunk 7]
```diff
@@ -41,31 +47,60 @@ protected function setUp(): void
...
+     * @param string $newItemValue
+     * @param string|null $orderCurrencyCode
+     * @dataProvider prepareDataSourceDataProvider
+     */
+    public function testPrepareDataSource(
+        $itemName,
+        $oldItemValue,
+        $newItemValue,
+        $orderCurrencyCode
+    ): void {
```
Comment (ihor-sviziev) [inline app/code/Magento/Sales/Test/Unit/Ui/Component/Listing/Column/PurchasedPriceTest.php]:
```suggestion
    public function testPrepareDataSource(
        string $itemName,
        string $oldItemValue,
        string $newItemValue,
        ?string $orderCurrencyCode
    ): void {
```

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
I still believe we can iterate all the orders and get all needed info with a single request instead of loading each order in the loop. Please update the solution

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

File: app/code/Magento/Sales/Ui/Component/Listing/Column/PurchasedPrice.php [hunk 8]
```diff
@@ -37,21 +49,30 @@ class PurchasedPrice extends Column
...
     public function __construct(
         ContextInterface $context,
         UiComponentFactory $uiComponentFactory,
         PriceCurrencyInterface $priceFormatter,
         array $components = [],
         array $data = [],
-        Currency $currency = null
+        Currency $currency = null,
+        OrderRepositoryInterface $orderRepository = null,
+        SearchCriteriaBuilder $searchCriteriaBuilder = null
```
Comment (zakdma) [inline app/code/Magento/Sales/Ui/Component/Listing/Column/PurchasedPrice.php]: It is better use SearchCriteriaBuilderFactory and create SearchCriteriaBuilder using it create() method

File: app/code/Magento/Sales/Ui/Component/Listing/Column/PurchasedPrice.php [hunk 9]
```diff
@@ -61,8 +82,10 @@ public function __construct(
     public function prepareDataSource(array $dataSource)
     {
         if (isset($dataSource['data']['items'])) {
+            $orderIds = array_column($dataSource['data']['items'],'order_id');
+            $orderCurrencyCodes = $this->getOrdersCurrency($orderIds);
```
Comment (zakdma) [inline app/code/Magento/Sales/Ui/Component/Listing/Column/PurchasedPrice.php]: This additional request to DB can fail PAT tests.
Is there another way to correctly get currency? Why it not appeared in $item['order_currency_code']?
Maybe make sense change request that fill dataSource?

**Review by zakdma — CHANGES_REQUESTED** [review]

**Comment by engcom-Echo** [comment]
@magento run all tests

File: app/code/Magento/Sales/Model/ResourceModel/Order/Creditmemo/Grid/Collection.php [hunk 10]
```diff
@@ -33,4 +33,32 @@ public function __construct(
...
+     */
+    protected function _translateCondition($field, $condition)
+    {
+        if ($field !== 'order_currency_code'
+            && !isset($this->_map['fields'][$field])
+        ) {
+            $this->_map['fields'][$field] = 'main_table.' . $field;
+        }
+
+        return parent::_translateCondition($field, $condition);
```
Comment (zakdma) [inline app/code/Magento/Sales/Model/ResourceModel/Order/Creditmemo/Grid/Collection.php:48]: I suppose parent call should be in front of your actions

File: app/code/Magento/Sales/Model/ResourceModel/Order/Creditmemo/Grid/Collection.php [hunk 11]
```diff
@@ -33,4 +33,32 @@ public function __construct(
...
+    }
+
+    /**
+     * @inheritDoc
+     */
+    protected function _renderFiltersBefore()
+    {
+        $this->getSelect()->join(
```
Comment (zakdma) [inline app/code/Magento/Sales/Model/ResourceModel/Order/Creditmemo/Grid/Collection.php]: please change to left join

File: app/code/Magento/Sales/Model/ResourceModel/Order/Creditmemo/Grid/Collection.php [hunk 12]
```diff
@@ -33,4 +33,32 @@ public function __construct(
...
+            ['cgf' => $this->getTable('sales_order_grid')],
+            'main_table.order_id = cgf.entity_id',
+            [
+                'order_currency_code'=>'order_currency_code'
```
Comment (zakdma) [inline app/code/Magento/Sales/Model/ResourceModel/Order/Creditmemo/Grid/Collection.php]: add spaces around =>
and add comma at the end

File: app/code/Magento/Sales/Model/ResourceModel/Order/Creditmemo/Order/Grid/Collection.php [hunk 13]
```diff
@@ -33,4 +33,32 @@ public function __construct(
...
+     */
+    protected function _translateCondition($field, $condition)
+    {
+        if ($field !== 'order_currency_code'
+            && !isset($this->_map['fields'][$field])
+        ) {
+            $this->_map['fields'][$field] = 'main_table.' . $field;
+        }
+
+        return parent::_translateCondition($field, $condition);
```
Comment (zakdma) [inline app/code/Magento/Sales/Model/ResourceModel/Order/Creditmemo/Order/Grid/Collection.php:48]: I suppose parent call should be in front of your actions

File: app/code/Magento/Sales/Model/ResourceModel/Order/Creditmemo/Order/Grid/Collection.php [hunk 14]
```diff
@@ -33,4 +33,32 @@ public function __construct(
...
+    /**
+     * @inheritDoc
+     */
+    protected function _renderFiltersBefore()
+    {
+        $this->getSelect()->join(
+            ['cgf' => $this->getTable('sales_order_grid')],
+            'main_table.order_id = cgf.entity_id',
+            [
+                'order_currency_code'=>'order_currency_code'
```
Comment (zakdma) [inline app/code/Magento/Sales/Model/ResourceModel/Order/Creditmemo/Order/Grid/Collection.php]: add spaces around =>
and add comma at the end

File: app/code/Magento/Sales/Model/ResourceModel/Order/Creditmemo/Order/Grid/Collection.php [hunk 15]
```diff
@@ -33,4 +33,32 @@ public function __construct(
...
+
+        return parent::_translateCondition($field, $condition);
+    }
+
+    /**
+     * @inheritDoc
+     */
+    protected function _renderFiltersBefore()
+    {
+        $this->getSelect()->join(
```
Comment (zakdma) [inline app/code/Magento/Sales/Model/ResourceModel/Order/Creditmemo/Order/Grid/Collection.php]: please change to left join

**Review by zakdma — CHANGES_REQUESTED** [review]

**Comment by engcom-Echo** [comment]
@magento run all tests

**Review by zakdma — APPROVED** [review]

**Comment by dmitriyprime** [comment]
@magento i am working on it

**Comment by zakdma** [comment]
@magento import code to https://github.com/magento-tsg/magento2ce

**Comment by magento-engcom-team** [comment]
@zakdma, an error occurred during the Pull Request import.

**Comment by magento-engcom-team** [comment]
@zakdma the branch with code successfully imported into`magento-tsg/magento2ce` repository. Branch name: `imported-magento-magento2-31553`.

**Comment by engcom-Bravo** [comment]
:heavy_check_mark: QA Passed

### Manual testing scenario

1. From Admin Panel, go to _Stores - All Stores_ and create a second Website, Store, Store View
2. Go to _Stores - Configuration_ and change the **Scope** for the previously created second Store View
3. Go to  _General - Currency Setup_ and add  _Ukrainian Hryvnia_ to **Allowed Currencies**
4. Change the **Default Display Currency** to _Ukrainian Hryvnia_. **Save Config**
5. Go to _Stores - Currency Rates_ and set _2_ for UAH. **Save Currency Rates**
6. Create an order from the second Store, where the currency is UAH
7. From Admin Panel, go to _Sales - Orders_. Create an **Invoice** for the previously created order
8. Create a **Credit Memo** for the previously created order
9. Open **Orders** Grid.
10. Click **Columns** and check _Subtotal_ and _Shipping and Handling_
11. Observe the Currency symbol in _Grand Total (Base)_, _Subtotal_ and _Shipping and Handling_ columns
The currency symbol is Dollar $
dollar
12. Go to _Sales - Credit Memos_ grid and observe Currency symbol in _Refund_, _Subtotal_ and _Shipping and Handling_ columns
The currency symbol is Dollar $
cr_m_dollar

**AFTER** applying changes provided in the PR the Currency symbol in the _Credit Memo_ grid is displayed correctly
after

after2

**Pull request merged**